### PR TITLE
Uncalled function checks don't work with negation

### LINKED ISSUE #43096: Uncalled function checks don't work with negation
# Bug Report

### 🔎 Search Terms

`uncalled function checks`, `negation`

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ssl=16&ssc=2&pln=1&pc=1#code/PTAEAEGcBcCcEsDG0ByBXANhgwgCwKaIDWkAXKHGvgFDUAmhGAhrPqAGZoB2y8A9l1DxIAMT58AFAEpyAI3EZ8TLgG56jFm0QCYQyAFUuDdvC746YvuQnTQAXgB8oeX0XKpoAD6hux0+bVqeHZQCWFLDwBvalBY0BAfHiYscw5uXgFQRAJiSFBk1iY6AE9QAHc+WCJy-CxQAlZqAF9aYNCAQnDxKJi4hPTkxTo0nmh+QWzCElBIXD5MYe0AN3xYClxhLKZINmhxZtaQiU6DI3wTMwtu0Gi4+LBoDbzEbbZZ+Yxhrj5oZy0+FasOgHIA)

### 💻 Code

```ts
declare function isFoo(): boolean;
declare const isUndefinedFoo: (() => boolean) | undefined;

if (isFoo) {
    // uncalled function checks already work well here
}

if (!isFoo) {
    // uncalled function checks should cover this case too
}

if (!isUndefinedFoo) {
    // this case should not be covered
}
```

### 🙁 Actual behavior

Uncalled function checks do not work with the `!` operator.

### 🙂 Expected behavior

Unless a function may be undefined it would be very helpful to check whether it is called when it is used with the `!` operator.

I created this issue to separate a part of the implementation of #42835 resulting from issue #35584. This specific issue was inspired by https://github.com/microsoft/TypeScript/issues/35584#issuecomment-769401594.

### PR DESCRIPTION
**Pull request by jonhue**
Fixes #43096

### CODE REVIEW

File: tests/baselines/reference/truthinessCallExpressionCoercion2.errors.txt [hunk 1]
```diff
@@ -87,8 +86,6 @@ tests/cases/compiler/truthinessCallExpressionCoercion2.ts(112,14): error TS2774:
         // error
         typeof window !== 'undefined' && window.console &&
             ((window.console as any).firebug || (window.console.exception && window.console.table));
-                                                 ~~~~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2774: This condition will always return true since the function is always defined. Did you mean to call it instead?
```
Comment (jonhue) [inline tests/baselines/reference/truthinessCallExpressionCoercion2.errors.txt:91]: see https://github.com/microsoft/TypeScript/pull/42835#discussion_r577538453

**Comment by sandersn** [comment]
I prefer to wait on a PR until after a bug is accepted because code goes stale so fast. That said, I believe there are a couple of related PRs that are ready to merge, so maybe it's worth deciding whether to accept the source issue for this too.

File: tests/cases/compiler/uncalledFunctionChecksDontWorkWithNegation.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    // error
+}
+
+if (!isFoo) {
+    // error
+}
+
+if (!isUndefinedFoo) {
+    // no error
+}
```
Comment (gabritto) [inline tests/cases/compiler/uncalledFunctionChecksDontWorkWithNegation.ts:16]: would be nice to have tests for when those negated expressions are part of larger logical expressions (e.g. `!isFoo || x`, etc)

Comment (gabritto) [inline tests/cases/compiler/uncalledFunctionChecksDontWorkWithNegation.ts:16]: would also like to see what happens when we issue this error for unawaited promises, like so:
```ts
declare function foo(): Promise<string | undefined>;
const awaitedValue = foo();
if (awaitedValue) { // Emits ts2801
    // ...
}
if (!awaitedValue) { // What happens here?
    // ...
}
```

**Comment by gabritto** [comment]
> I prefer to wait on a PR until after a bug is accepted because code goes stale so fast. That said, I believe there are a couple of related PRs that are ready to merge, so maybe it's worth deciding whether to accept the source issue for this too.
> 
> Edit: #42835

At a first glance, I think we should accept the issue. Before a final decision though, we should user/DT test this to see if we get too many unwanted errors.

**Comment by jonhue** [comment]
Hey @gabritto, thanks for following up on this!

To be completely honest, it's been quite a while and I have a bunch of other projects at hand right now. I will try to squeeze fixing the merge conflicts into my schedule in the next few days, but to be honest, I'm not too sure i'll manage to.

So feel free to take over the pr if you think this change is still worth adding :)

**Comment by sandersn** [comment]
@gabritto Do you mind if I close this until you have a chance to get back to it?

**Comment by gabritto** [comment]
> @gabritto Do you mind if I close this until you have a chance to get back to it?

no, go ahead

**Pull request closed without merging**